### PR TITLE
Added LogGamma and tests

### PR DESCRIPTION
**Pull request by abhinavagarwal07**

### CODE REVIEW

File: symengine/functions.h [hunk 1]
```diff
@@ -1010,6 +1010,37 @@ class UpperGamma: public Function {
...
+     * \return whether the 2 objects are equal
+     * */
+    virtual bool __eq__(const Basic &o) const;
+    virtual int compare(const Basic &o) const;
+    //! \return Size of the hash
+    virtual std::size_t __hash__() const;
+    //! \return `true` if canonical
+    bool is_canonical(const RCP<const Basic> &arg) const;
+    virtual vec_basic get_args() const { return {arg_}; }
+    RCP<const Basic> rewrite_as_intractable() const;
```
Comment (isuruf) [inline symengine/functions.h]: `rewrite_as_gamma` is better here

**Comment by isuruf** [comment]
Can you add a `subs` method and most importantly evaluations?

**Comment by abhinavagarwal07** [comment]
sure.

File: symengine/eval_arb.cpp [hunk 2]
```diff
@@ -320,9 +320,15 @@ class EvalArbVisitor : public BaseVisitor<EvalArbVisitor> {
...
-    void bvisit(const Gamma &) {
-        throw std::runtime_error("Not implemented.");
+    void bvisit(const Gamma &x) {
+        apply(result_, *(x.get_args())[0]);
+        arb_gamma(result_, result_, prec_);
     };
+    void bvisit(const LogGamma &x){
+        apply(result_, *(x.get_args())[0]);
+        arb_gamma(result_, result_, prec_);
+        arb_log(result_, result_, prec_);
```
Comment (isuruf) [inline symengine/eval_arb.cpp]: Use `arb_lgamma` as precision of `log(gamma())` is lesser.

File: symengine/eval_mpfr.cpp [hunk 3]
```diff
@@ -224,6 +224,12 @@ class EvalMPFRVisitor : public BaseVisitor<EvalMPFRVisitor> {
         mpfr_gamma(result_, result_, rnd_);
     };
 
+    void bvisit(const LogGamma& x) {
+        apply(result_, *(x.get_args()[0]));
+        mpfr_gamma(result_, result_, rnd_);
+        mpfr_log(result_, result_, rnd_);
```
Comment (isuruf) [inline symengine/eval_mpfr.cpp]: Use `mpfr_lngamma`

File: symengine/eval_mpc.cpp [hunk 4]
```diff
@@ -265,6 +265,23 @@ class EvalMPCVisitor : public BaseVisitor<EvalMPCVisitor> {
         }
     }
 
+    void bvisit(const Gamma &x) {
```
Comment (isuruf) [inline symengine/eval_mpc.cpp]: Argument of `x` can be complex

Comment (abhinavagarwal07) [inline symengine/eval_mpc.cpp]: mpc doesn't has `gamma` implementation. So I guess, this should throw error for now.

**Comment by isuruf** [comment]
Can you add loggamma to `eval_double` as well?

**Comment by isuruf** [comment]
Also to `printer.cpp`

File: symengine/tests/eval/test_eval_double.cpp [hunk 5]
```diff
@@ -81,6 +82,11 @@ TEST_CASE("eval_double: eval_double", "[eval_double]")
         { mul(EulerGamma, integer(10)), 5.7721566490153286 },
         { max({r2, r1}), 0.841470984808},
         { min({add(r1, r4), r2}), 0.479425538604},
+        { gamma(integer(2)), 1 },
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_double.cpp]: Isn't `gamma(2)` automatically simplified to `1`?

**Comment by isuruf** [comment]
Sorry to bother you again. Can you add loggamma to `lambda_double` as well?

File: symengine/tests/eval/test_eval_double.cpp (hunk 5 above)
Comment (abhinavagarwal07) [inline symengine/tests/eval/test_eval_double.cpp]: yes it is
 `gamma(4)` also automatically simplifies to `6`

Comment (isuruf) [inline symengine/tests/eval/test_eval_double.cpp]: Then you should add a test that actually evaluates a `Gamma` instance. Here you are evaluating a `Integer` object

File: symengine/tests/eval/test_eval_mpfr.cpp [hunk 6]
```diff
@@ -77,5 +78,15 @@ TEST_CASE("precision: eval_mpfr", "[eval_mpfr]")
     REQUIRE(mpfr_cmp_d(a, 2.00000000000001) == -1);
     REQUIRE(mpfr_cmp_d(a, 1.99999999999999) == 1);
 
+    r = loggamma(integer(3));
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_mpfr.cpp]: This is automatically simplified right? So, this is not evaluating `loggamma(3)`, but `log(gamma(3))`

File: symengine/tests/eval/test_eval_mpfr.cpp [hunk 7]
```diff
@@ -77,5 +78,15 @@ TEST_CASE("precision: eval_mpfr", "[eval_mpfr]")
     REQUIRE(mpfr_cmp_d(a, 2.00000000000001) == -1);
     REQUIRE(mpfr_cmp_d(a, 1.99999999999999) == 1);
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_mpfr.cpp]: Unrelated to the PR. Same situation as before. `r` is evaluated to `2` and `min` is not simplified. Better test case would have been `min(sqrt(2), 2)`

File: symengine/tests/eval/test_eval_double.cpp [hunk 8]
```diff
@@ -81,6 +82,9 @@ TEST_CASE("eval_double: eval_double", "[eval_double]")
         { mul(EulerGamma, integer(10)), 5.7721566490153286 },
         { max({r2, r1}), 0.841470984808},
         { min({add(r1, r4), r2}), 0.479425538604},
+        { gamma(div(integer(4), integer(3))), 0.892979511569249211},
+        { loggamma(div(integer(7), integer(2))), 1.200973602347074224 },
+        { loggamma(integer(3)), 0.693147180559945309 },
```
Comment (isuruf) [inline symengine/tests/eval/test_eval_double.cpp]: Same issue here.

File: symengine/functions.cpp [hunk 9]
```diff
@@ -3051,6 +3051,79 @@ RCP<const Basic> uppergamma(const RCP<const Basic> &s, const RCP<const Basic> &x
...
+    if (arg == arg_)
+        return rcp_from_this();
+    else
+        return loggamma(arg);
+}
+
+RCP<const Basic> loggamma(const RCP<const Basic> &arg)
+{
+    if (is_a_Number(*arg) and not
+        static_cast<const Number &>(*arg).is_positive()) {
```
Comment (isuruf) [inline symengine/functions.cpp]: This should check whether arg is an integer or not. loggamma has a pole only at negative integers

Comment (abhinavagarwal07) [inline symengine/functions.cpp]: `loggamma(0) = loggamma(-ve) = infinity `
http://www.wolframalpha.com/input/?i=loggamma(0)

Comment (isuruf) [inline symengine/functions.cpp]: Yes, it should be loggamma has a pole only at non-positive integers.
http://www.wolframalpha.com/input/?i=loggamma(-1%2F2)

Comment (isuruf) [inline symengine/functions.cpp]: If you can make the change to non-positive integers, then we can get this merged

Comment (abhinavagarwal07) [inline symengine/functions.cpp]: Sorry, got busy with assignments and classes. 
Will do it.

**Comment by isuruf** [comment]
+1 to merge if tests pass

**Pull request merged**